pkg/rpm/database.go

```diff
@@ -0,0 +1,285 @@
+		rpmStdout:  stdoutPipe,
+		rpmScanner: bufio.NewScanner(stdoutPipe),
+	}
+}
+
+// ProvidesFile return true if a package provides the specified path in the RPM
+// database. If the path is relative, we assume its relative to the root
+// directory.
+func (d *rpmDatabase) ProvidesFile(path string) bool {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`--query`,
+		`--whatprovides`, path,
+	)
+	if err := cmd.Run(); err != nil {
+		// When rpm does not provide a file, the expected exit status is 1. On non-zero
+		// we always default to returning the file is not provided, but anything other
+		// than 1 is considered unexpected.
+		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() != 1 {
+			logrus.WithError(err).Errorf(
+				"unexpected exit status when querying %s is provided by an RPM package", path)
+		}
+		return false
+	}
+	// The rpm exited properly, which implies the file IS provided by an RPM
+	// package.
+	return true
+}
+
+// Delete removes all the RPM database files.
+func (d *rpmDatabase) Delete() error {
+	return os.RemoveAll(d.dbPath)
+}
+
+// Next retrieves the next item in the query and make it available through the Package call.
+func (q *rpmDatabaseQuery) Next() bool {
+	if q.err != nil || q.rpmScanStopped {
+		return false
+	}
+	q.nextPackage = rpmPackage{}
+	for i := 0; q.rpmScanner.Scan(); i++ {
+		line := strings.TrimSpace(q.rpmScanner.Text())
+		if line == "" || strings.HasPrefix(line, "(none)") {
+			continue
+		}
+		if line == "." {
+			// Reached package delimiter. Ensure the current package is well-formed. If it
+			// is, add it to the return slice.
+			if q.nextPackage.Name != "" && q.nextPackage.Version != "" && q.nextPackage.Arch != "" {
+				return true
+			}
+			// Start a new package definition and reset 'i'.
+			q.nextPackage = rpmPackage{}
+			i = -1
+			continue
+		}
+		switch i {
+		case 0:
+			// This is not a real package. Skip it...
+			if line == "gpg-pubkey" {
+				continue
+			}
+			q.nextPackage.Name = line
+		case 1:
+			q.nextPackage.Version = line
+		case 2:
+			q.nextPackage.Arch = line
+		case 3:
+			moduleSplit := strings.Split(line, ":")
+			if len(moduleSplit) < 2 {
+				continue
+			}
+			moduleStream := fmt.Sprintf("%s:%s", moduleSplit[0], moduleSplit[1])
+			q.nextPackage.Module = moduleStream
+		default:
+			// i >= 4 is reserved for provided filenames.
+			q.nextPackage.Filenames = append(q.nextPackage.Filenames, line)
+		}
+	}
+	q.rpmScanStopped = true
+	// If we stopped due to pipe errors, parse it and return.
+	if err := q.rpmScanner.Err(); err != nil {
+		if q.rpmStderr.Len() != 0 {
+			logrus.Warnf("Error executing RPM rpm: %s", q.rpmStderr.String())
+		}
+		q.err = errors.Errorf("rpm: error reading rpm output: %v", err)
+		utils.IgnoreError(q.rpmStdout.Close)
+		return false
+	}
+	// Otherwise, let's Wait() to and set the rpm command status.
+	q.err = q.rpmWait()
```

Comment:
I believe we end up here if there aren't anymore packages and there were no errors. In that case, I think `q.rpmStdout.Close` is still open, and we need to make sure we close it here, too

Reply:
[Wait() closes all the sub-process file descriptors](https://pkg.go.dev/os/exec#Cmd.StdoutPipe)